### DISCUSSION TITLE
How could i get a permanent link for raw file?

### DISCUSSION THREAD

**Discussion by Lotaristo** [community/community#22537] [New to GitHub]
<p>I am searching google for about 3 hours and stil not succeed in this.<br>
I do data analysis in Python, and i’ve seen a lot of tymes that people unstead of downloading file from hard drive just paste raw link from github directory. If I get this link (clicking ‘raw’ in ma repository), I get a link which will expire in 10 minutes. So, how could i get a permanent link (or at least for a month or so)?<br>
I’ve seen answers that i have to do it using API and get user token, but i’m pretty bad with it and no matter how i tried i didn’t managed to get what i want. So, I’ll be really glad if could help me…</p>

**Comment by byrneh** [comment, marked as answer]
<p>HI <a class="mention" href="/lotaristo">@Lotaristo</a>,</p>
<p>a worked example for you using a PUBLIC repository, please note you SHOULD use the same URL format for a PRVATE repository and cannot use the short lived expiring token shown in the browser session passed as a query parameter but will need to create a personal access token with permissions.</p>
<p>GitHub file in browser</p>
<blockquote>
<p><a href="https://github.com/octocat/Hello-World/blob/master/README" rel="noopener nofollow ugc">https://github.com/octocat/Hello-World/blob/7fd1a60b01f91b314f59955a4e4d4e80d8edf11d/README</a></p>
</blockquote>
<p>RAW format URL</p>
<blockquote>
<p><a href="https://raw.githubusercontent.com/octocat/Hello-World/master/README" rel="noopener nofollow ugc">https://raw.githubusercontent.com/octocat/Hello-World/master/README</a></p>
</blockquote>
<p>If your repository file is a PRIVATE visibility repository you will need to pass a <strong>personal access token</strong> with scope permission to READ from the repository in an authentication header by adding</p>
<blockquote>
<p>-H “Authorization: Token $MY_PAT”</p>
</blockquote>
<p>If unsure how to get a personal access token see this link <a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token" rel="noopener nofollow ugc">creating-a-personal-access-token</a>, In my example variable $MT+PAT holds the personal teh value of my personal access token previously created</p>
<p>so you full format will like</p>
<blockquote>
<p><a href="https://github.com/octocat/Hello-World/blob/master/README" rel="noopener nofollow ugc">https://github.com/octocat/Hello-World/blob/7fd1a60b01f91b314f59955a4e4d4e80d8edf11d/README</a> -H “Authorization: Token $MY_PAT”</p>
</blockquote>
<p>This will be the solution to your question.</p>

**Comment by Lotaristo** [comment]
<p>Well, public worked well for me, thanks!</p>

**Comment by KengKoh-Sifive** [comment]
is there a way for private repo?

**Reply by DevBadger10** [reply]
You need to follow the instructions in the Selected Answer.

> HI [@Lotaristo](/lotaristo),
> 
> a worked example for you using a PUBLIC repository, please note you SHOULD use the same URL format for a PRVATE repository and cannot use the short lived expiring token shown in the browser session passed as a query parameter but will need to create a personal access token with permissions.
> 
> GitHub file in browser
> 
> > [https://github.com/octocat/Hello-World/blob/7fd1a60b01f91b314f59955a4e4d4e80d8edf11d/README](https://github.com/octocat/Hello-World/blob/master/README)
> 
> RAW format URL
> 
> > https://raw.githubusercontent.com/octocat/Hello-World/master/README
> 
> If your repository file is a PRIVATE visibility repository you will need to pass a **personal access token** with scope permission to READ from the repository in an authentication header by adding
> 
> > -H “Authorization: Token $MY_PAT”
> 
> If unsure how to get a personal access token see this link [creating-a-personal-access-token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token), In my example variable $MT+PAT holds the personal teh value of my personal access token previously created
> 
> so you full format will like
> 
> > [https://github.com/octocat/Hello-World/blob/7fd1a60b01f91b314f59955a4e4d4e80d8edf11d/README](https://github.com/octocat/Hello-World/blob/master/README) -H “Authorization: Token $MY_PAT”
> 
> This will be the solution to your question.

**Reply by magendirans** [reply]
Thanks @DevBadger10,

It's working fine for me.

**Reply by eldepor** [reply]
That instructions is actually missing exactly what the OP is asking for: the RAW format URL of a file from a private repository but whose token does not expire. Where is it? He wrote it in the public repo example, but for the private one he only wrote the "github url file in browser" format. Thanks.

**Comment by nickjuntilla** [comment]
Doers not work even for public files, just a 404 error.

**Reply by DevBadger10** [reply]
You can do public by removing the token.

**Reply by copernico** [reply]
I tried removing the token part from the URL, it does not work.

**Comment by nickjuntilla** [comment]
If you remove /blob it works

**Reply by DevBadger10** [reply]
But it'd be short lived.

**Comment by Satoru-sama** [comment]
> I am searching google for about 3 hours and stil not succeed in this. I do data analysis in Python, and i’ve seen a lot of tymes that people unstead of downloading file from hard drive just paste raw link from github directory. If I get this link (clicking ‘raw’ in ma repository), I get a link which will expire in 10 minutes. So, how could i get a permanent link (or at least for a month or so)? I’ve seen answers that i have to do it using API and get user token, but i’m pretty bad with it and no matter how i tried i didn’t managed to get what i want. So, I’ll be really glad if could help me…

**Reply by shubhamchopade** [reply]
Any luck figuring this out?

**Comment by tyconstack** [comment]
I'm needing help with this exact issue. For private, the raw URL keeps changing. I'd need something like a static raw URL.

**Comment by kalungia** [comment]
Hello, has anybody found a soultion for this

**Reply by ntriisii** [reply]
use the same format as the public repo , just add a new authorization header as the answer says for ex if you use curl the url would be :
`curl https://raw.githubusercontent.com/<repo-name>/<branch-name>/path/to/file -H "Authorization: Token <your-token>"`

**Comment by vkandala-ssquare** [comment]
For those who are still looking for the answer, hope this helps.

Since the tokens that are used to access raw downloads rotate regularly for security you'll need to authenticate your requests.
 
I'd recommend using the [get repository content](https://docs.github.com/rest/repos/contents?apiVersion=2022-11-28#get-repository-content) REST API endpoint with a [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens). This will return a json object with a fresh download_url each time that will retrieve the file.

curl -L \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer [TOKEN]" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/[organization]/[repository]/contents/[myFile] | jq '.download_url' | xargs curl -O

Where '[organization]' can be your enterprise organization handle or you can replace it with your personal handle and '[repository]' is the repository name where the file is stored.

'[TOKEN]' - is your personal access token.  After you have created it you have to provide authorization to be used against your organization, if needed.

You can try using Postman to test the GET API call for retrieving the contents of the file and the 'download_url'.

Thanks,
Vamsi

**Comment by apakhomov** [comment]
PAT is not a good solution since it gives access to all files in repo, but I need to give access to just one file (considered non-confidential) by constant URL. Even fine-grained tokens do not provide any mechanism to restrict access by file or path.

**Comment by Midas-Coin** [comment]
This is disappointing; raw should be exactly that, raw...

**Comment by tyleratverkada** [comment]
I was expecting the .netrc to put in my token where it's needed. But the Authorization header needs to be in place. For this, I need scripting. :-/

**Comment by Naryalin** [comment]
I am facing the same issue with a private repository. In my case, I have a portfolio that retrieves data from the GitHub API, and I'm able to successfully fetch the README, about section, languages, etc. The main issue arises when dealing with images, such as logos or mockups, that are stored in the repositories.

Using the GitHub API, I can retrieve a `download_url` for these images, which works well initially. However, in production, because the page is served multiple times, the images eventually stop being displayed due to the `download_url `expiration.

Here's a sample of the code I'm using:
```typescript
const octokit = new Octokit({
    auth: getSecret("GITHUB_ACCESS_TOKEN"),
})
async function getRepoAsset(repoUrl:string, assetUrl: string) {
  try {
    const { data } = await octokit.request(`GET ${repoUrl}/contents/${assetUrl}`, {
      owner: 'OWNER',
      repo: 'REPO',
      path: 'PATH',
      headers: {
        'accept': 'application/vnd.github.object+json',
        'X-GitHub-Api-Version': '2022-11-28'
      }
    })
    
    return data.download_url;
  } catch (error) {
    console.error(error);
  }
}
```

Is there a way to obtain a URL that doesn't expire or another solution to serve these images indefinitely in a production environment, **without downloading and saving the images**? Doing so defeats the purpose of having the images in the repository, as I might as well just store the images in services like S3 and store those URLs directly in the README.

Thanks in advance for any advice you can share!

**Comment by Decky281268** [comment]
Jayavision

**Comment by pisoni-onshape** [comment]
This is so amazingly stupid. I need ONE STATIC RAW URL to a non-private file in my **private** repository (to a fixed commit - so technically it's always constant) to someone without revealing my tokens OR without worrying about expiring times. Why is that not possible!! Why did no one at Github think of that?

**Comment by JonasDoe** [comment]
I want to refer a a remote opencode instruction from a private repo (https://opencode.ai/docs/rules/#custom-instructions), and for this I'ld need a permanent url. I have no options to add an `Authorization` header etc.
